src/iterators/bfs.jl

```diff
@@ -0,0 +1,95 @@
+"""
+    BFSIterator
+
+`BFSIterator` is a subtype of [`VertexIterator`](@ref) to iterate through graph vertices using a breadth-first search. A source node(s) is optionally supplied as an `Int` or an array-like type that can be indexed if supplying multiple sources. If no source is provided, it defaults to the first vertex.
+    
+# Examples
+```julia-repl
+julia> g = smallgraph(:house)
+{5, 6} undirected simple Int64 graph
+
+julia> for node in BFSIterator(g,3)
+           display(node)
+       end
+3
+1
+4
+5
+2
+```
+"""
+struct BFSIterator{S,G<:AbstractGraph}
+    graph::G
+    source::S
+end
+
+"""
+    BFSVertexIteratorState
+
+`BFSVertexIteratorState` is a struct to hold the current state of iteration
+in BFS which is needed for the `Base.iterate()` function.
+"""
+mutable struct BFSVertexIteratorState
+    visited::BitVector
+    queue::Vector{Int}
+    neighbor_idx::Int
+    n_visited::Int
+end
+
+BFSIterator(g::AbstractGraph) = BFSIterator(g, first(vertices(g)))
+
+Base.length(t::BFSIterator) = nv(t.graph)
```

Comment:
I added a known length but on second thought it may not work for unconnected graphs. In which case the termination criteria in the iteration also need to change

Reply:
yes, we need to assume an unknown length in all cases I guess

Reply:
yeah you're right sorry